### PR TITLE
Various work for upstream

### PR DESCRIPTION
**Pull request by madebr**
brender changes:
- Add a few function declarations to `brender.h`, such that dethrace may use it.
- change the typedef for `br_size_t` from `uint32_t` to `size_t`.
- sort the tokens in `br_types.h` by value, instead of by name.
- implement all genclip functions (=clipping of rectangles with rectangles + points)
- implement brender  lexer, associative array, dbsetup, resource, resreg, pmnull functions
- Fix the brender init string. It is different between the DOS executable and Windows
- Use `BR_ASIZE` and `COUNT_OF` in various places instead of hardcoding magic values

dethrace changes:
- implement all missing input functions
- implement various edit mode callback functions
- Add `WARN_ONCE` macro to avoid spam
- implement dimmed rounded rectangled for the HUD background

various:
- add option to select orderkey in `tools/progress.py`, by ordering on name it becomes easy to diff progress
- add script to decode text files (`tools/decode_datatxt.py`)

plus many other changes

```
name                                          percent	completed / total
=======================================
total                                            46%	1609 / 3467
```

The big bump in percentage is due to implementing many little functions in `powerup.c` and `pmnull.c` + others.
TBH, this percentage says nothing about real progress.

Screenshot from 2022-02-22 20-32-35

Screenshot from 2022-02-22 22-01-59

Screenshot from 2022-02-22 22-02-08

### CODE REVIEW

File: src/BRSRC13/CORE/FW/resource.c [hunk 1]
```diff
@@ -1,32 +1,44 @@
...
     LOG_TRACE9("(%p)", r);
 
-    p = r;
-    p = p - sizeof(resource_header);
-    if (((resource_header*)p)->magic_num != 0xDEADBEEF) {
+    p = (br_uint_8*)r - 1;
+    while (*p == 0) {
+        p--;
+    }
+    fflush(stdout);
```
Comment (madebr) [inline src/BRSRC13/CORE/FW/resource.c:37]:
```suggestion
```

File: src/DETHRACE/common/loading.c [hunk 2]
```diff
@@ -3138,15 +3219,78 @@ void ReadNetworkSettings(FILE* pF, tNet_game_options* pOptions) {
...
+    BAIL_IF_NEGATIVE(PrintNetOptions(f, 0));
+    BAIL_IF_NEGATIVE(PrintNetOptions(f, 1));
+    BAIL_IF_NEGATIVE(PrintNetOptions(f, 2));
+    BAIL_IF_NEGATIVE(PrintNetOptions(f, 3));
+    BAIL_IF_NEGATIVE(PrintNetOptions(f, 4));
+    BAIL_IF_NEGATIVE(PrintNetOptions(f, 5));
+    BAIL_IF_NEGATIVE(PrintNetOptions(f, 6));
+    BAIL_IF_NEGATIVE(PrintNetOptions(f, 7));
+    
+    fclose(f);
```
Comment (madebr) [inline src/DETHRACE/common/loading.c:3291]:
```suggestion
    BAIL_IF_NEGATIVE(PrintNetOptions(f, 7));

#undef BAIL_IF_NEGATIVE

    fclose(f);
```

File: src/DETHRACE/common/finteray.c [hunk 3]
```diff
@@ -178,6 +178,7 @@ int DRSceneRayPick2D(br_actor* world, br_vector3* pPosition, br_vector3* pDir, d
     LOG_TRACE("(%p, %p, %p, %p)", world, pPosition, pDir, callback);
 
     BrMatrix34Inverse(&gPick_model_to_view__finteray, &world->t.t.mat);
+    LOG_WARN_ONCE("Missing material and model pointers to ActorRayPick2D");
```
Comment (dethrace-labs) [inline src/DETHRACE/common/finteray.c:181]: Whats the purpose of the warning here? These params are used as defaults and are always deliberately set to null

File: src/DETHRACE/common/init.c [hunk 4]
```diff
@@ -608,6 +610,7 @@ void InitRace() {
     InitPlayers();
     if (gNet_mode) {
         gProgram_state.credits_earned = gInitial_net_credits[gCurrent_net_game->options.starting_money_index];
+        InitNetGameplayStuff();
```
Comment (dethrace-labs) [inline src/DETHRACE/common/init.c:613]: 👍

File: src/DETHRACE/common/input.c [hunk 5]
```diff
@@ -520,20 +558,59 @@ void ChangeTextTo(int pXcoord, int pYcoord, char* pNew_str, char* pOld_str) {
...
+                if (new_char == ' ') {
+                    let->letters[0] = ' ';
+                }
+                let->current_offset = let->number_of_letters * gCurrent_graf_data->save_slot_letter_height;
+                let->rolling_type = new_type;
+            }
+        }
+    }
+    for (i = len; i < len2; i++) {
+        // new_type should be calculated here.
```
Comment (dethrace-labs) [inline src/DETHRACE/common/input.c:593]: is this a `TODO:` or a description of the behavior?

**Comment by dethrace-labs** [comment]
Amazing work!

File: src/DETHRACE/common/finteray.c (hunk 3 above)
Comment (madebr) [reply src/DETHRACE/common/finteray.c:181]: I think this function is called with a material and model pointer, but I don't know what one and how it is used.

File: src/DETHRACE/common/input.c (hunk 5 above)
Comment (madebr) [reply src/DETHRACE/common/input.c:593]: This comment is out of date and can be removed.

Comment (madebr) [reply src/DETHRACE/common/input.c:593]:
```suggestion
```

**Pull request merged**